copy/single.go

```diff
@@ -398,6 +400,29 @@ func (ic *imageCopier) compareImageDestinationManifestEqual(ctx context.Context,
 	}, nil
 }
 
+// canSkipValidation returns true if the source supports storing partial images and the image is.
+func canSkipValidation(ctx context.Context, annotations map[string]string, src private.ImageSource) (bool, error) {
+	toc, err := toc.GetTOCDigest(annotations)
+	if err != nil {
+		return false, err
+	}
+	if toc == nil {
+		return false, nil
+	}
```

Comment:
There’s a whole another question here about what it means to have signature verification enforced for such a “trusted” source transport. Do we just give up and let it through? Refuse to validate policies that require signatures (somehow)?

It seems a bit incongruous to let things through here, while `ic.canSubstituteBlobs` tries so hard to only sign data we have seen and verified. Admittedly c/storage is already a bit of a special case, and signing local in-storage images is probably extremely rare.

Or should we, at this path, actually require the transport not just to _declare_ that the blob is trusted, but to positively validate the full blob against the TOC? (Ugh… that would be a lot of critical extra code. Or does that already exist?)

Reply:
I've not looked into it, but if signature verification is enforced but I don't think we can use the partial pull feature at the moment.

We can skip the validation only if `canSubstituteBlobs`, I'll fix that.

Reply:
Oh … a random late-night thought: I should not be overcomplicating this.

The digests we validate come from `LayerInfosForCopy`, which is already transport-authored and not authenticated by the signed manifest. So there is a clear precedent for still considering signatures valid in that case.

---

Now, in that case we do validate the uncompressed digest… against data that was previously recorded, i.e. that protects against corruption but not malice. So validating the layer against the TOC would be consistent with that, OTOH I’m not sure there is a very good reason to do that — it makes sense if the policy + trusted keys + executables of Podman etc. were more trusted than the c/storage data. 

---

(On a push, don’t we need to discard the TOC and build one anew while compressing, anyway?)

Reply:
The TOC digest is used only in the image manifest (the `io.github.containers.zstd-chunked.manifest-checksum` annotation), so a different TOC will affect only the metadata and the compressed stream digest, it won't change the diffID.

I've pushed a new version, but I keep it marked as Draft for now as I want to experiment with this a bit more

Reply:
Overall, we went with `LayerInfosForCopy` returning (expected, untrusted) uncompressed digests for the TOC-pulled layers.

---

> (On a push, don’t we need to discard the TOC and build one anew while compressing, anyway?)

We do discard them AFAICS: `storageImageSource.LayerInfosForCopy` creates new `BlobInfo` data for each layer, and that does not preserve _any_ annotations. So it also discards the TOC annotations.